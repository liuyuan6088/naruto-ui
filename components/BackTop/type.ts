
import { ICommonProps } from '../utils/type';

export interface IBackTopProps extends ICommonProps {
  // 滚动高度达到此参数值才出现 BackTop  默认400
  visibilityHeight?: number;
  // 设置需要监听其滚动事件的元素，值为一个返回对应 DOM 元素的函数
  target?: () => (HTMLElement | Window);
  // 点击按钮的回调函数
  onClick?: (e: React.MouseEvent) => void;
}