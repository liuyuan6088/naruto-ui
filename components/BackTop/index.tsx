import * as React from 'react'
import { IBackTopProps } from './type'
import Animate from 'rc-animate'
import { throttle } from 'loadsh'
import Icon from '../Icon'
import cx from 'classnames'
import { scrollToY } from '../utils/scrollTo'
import { primaryName } from '../utils/constant'
import './style/index.less'

const { useState, useEffect } = React

const defaultProps: IBackTopProps = {
  prefixCls: `${primaryName}-backTop`,
  visibilityHeight: 400,
  target: () => window
}

const handleScrollTo = ({ onClick }: IBackTopProps, targetNode: HTMLElement | Window) => (
  e: React.MouseEvent
) => {
  targetNode === window ? scrollToY(0) : scrollToY(0, targetNode as HTMLElement)
  if (onClick) {
    onClick(e)
  }
}

const renderChildren = ({ prefixCls }: IBackTopProps, children: React.ReactNode) => {
  if (typeof children === 'string') {
    return <div className={`${prefixCls}-string`}>{children}</div>
  }
  if (children) {
    return children
  }
  return (
    <div className={`${prefixCls}-content`}>
      <Icon type="totop" color="#fff" />
    </div>
  )
}

const BackTop: React.FC<IBackTopProps> = props => {
  const { className, prefixCls, style, visibilityHeight, target, children } = props

  const [visible, setVisible] = useState<boolean>(false)

  const targetNode = target() || window
  const handleScroll = () => {
    setVisible(window.scrollY >= visibilityHeight)
  }

  useEffect(() => {
    const scroll = throttle(handleScroll, 40)
    targetNode.addEventListener('scroll', scroll)
    return () => {
      targetNode.removeEventListener('scroll', scroll)
    }
  }, [])

  const classes = cx(prefixCls, className)
  const backTopNode = (
    <div className={classes} style={style} onClick={handleScrollTo(props, targetNode)}>
      {renderChildren(props, children)}
    </div>
  )

  // 小于0 默认一直显示 解决<Animate>单侧不通过问题
  if (visibilityHeight < 0) {
    return backTopNode
  }

  return (
    <Animate component="" transitionName="fade">
      {visible ? backTopNode : null}
    </Animate>
  )
}

BackTop.defaultProps = defaultProps

export default BackTop
