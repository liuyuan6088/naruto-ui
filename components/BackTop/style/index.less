@import '../../styles/theme/default.less';
@import '../../styles/mixins/index.less';

@backTop-prefix-cls: ~"@{primary-name}-backTop";
@icon-prefix-cls: ~"@{primary-name}-icon";

.@{backTop-prefix-cls} {
  height: 40px;
  width: 40px;
  overflow: hidden;
  color: #fff;
  text-align: center;
  position: fixed;
  bottom: 50px;
  right: 50px;

  &-content {
    background-color: rgba(0,0,0,0.45);
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
    transition: all .3s @ease-in-out;
    &:hover {
      background-color: rgba(0,0,0,0.65);
    }
    > .@{icon-prefix-cls} {
      font-size: @font-size-large-0
    }
  }

  &-string {
    height: 100%;
    line-height: 40px;
    color: #fff;
    border-radius: @bordor-radius;
    background-color: @primary-color;
  }

  cursor: pointer;
}
