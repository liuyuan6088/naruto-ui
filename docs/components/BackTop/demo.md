---
imports:
  import Demo from './demo.tsx';
  import DemoCode from '!raw-loader!./demo.tsx';

---

## BackTop 回到顶部

用来代表用户或事物，支持图片、图标或字符展示。

## 使用BackTop

:::demo
  <CodeBox code={DemoCode}>
    <Demo />
  </CodeBox>
:::

## Props
| 参数     | 说明     | 类型   | 可选值 | 默认值 |
| -------- | -------- | ------ | ------ | ------ |
| visibilityHeight | 滚动高度达到此参数值才出现 | number | - | 400 |
| target | 设置需要监听其滚动事件的元素，值为一个返回对应 DOM 元素的函数 | () => (HTMLElement/Window) | - | () => (window) |
| onClick | 点击按钮的回调函数 | (e: React.MouseEvent) => void | - | - |



