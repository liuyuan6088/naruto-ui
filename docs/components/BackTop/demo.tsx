import * as React from 'react';
import { BackTop } from 'components';

const Demo: React.FC = () => {

  return (
    <React.Fragment>

      <p>简单使用</p>
      <p>可以自定义回到顶部按钮的样式</p>
      <p>有默认样式，距离底部 50px，可覆盖</p>
      <BackTop visibilityHeight={100} />
      <BackTop visibilityHeight={100} style={{ bottom: 100 }}>UP</BackTop>
    </React.Fragment>
  )
}

export default Demo;
