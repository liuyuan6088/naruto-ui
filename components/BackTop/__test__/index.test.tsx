import * as React from 'react'
import { render, mount } from 'enzyme'
import BackTop from '../index'

describe('render backTop', () => {
  it('render default', () => {
    const wrapper = render(<BackTop visibilityHeight={-1} />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('click BackTop', () => {
  it('should scroll to top after click it', async () => {
    const wrapper = mount(<BackTop visibilityHeight={-1} />)
    document.documentElement.scrollTop = 400
    wrapper.find('.naruto-backTop').simulate('click')
    await new Promise(resolve => setTimeout(resolve, 1000))
    expect(Math.abs(Math.round(document.documentElement.scrollTop))).toBe(0)
  })
})
