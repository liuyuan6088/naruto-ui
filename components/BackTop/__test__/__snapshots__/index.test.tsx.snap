// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render backTop render default 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "naruto-backTop",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "naruto-backTop-content",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "class": "naruto-icon naruto-icon-totop",
              "style": "color:#fff",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "href": "#icon-totop",
                },
                "children": Array [],
                "name": "use",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "href": "http://www.w3.org/1999/xlink",
                },
                "x-attribsPrefix": Object {
                  "href": "xlink",
                },
              },
            ],
            "name": "svg",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "style": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
